internal/xresource/unstructured/composite/composite.go

```diff
@@ -0,0 +1,420 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package composite contains an unstructured composite resource.
+package composite
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/ptr"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+
+	"github.com/crossplane/crossplane/internal/xresource/unstructured/reference"
+)
+
+// Schema specifies the schema version of a composite resource's Crossplane
+// machinery fields.
+type Schema int
+
+const (
+	// SchemaModern indicates a modern Namespaced or Cluster scope composite
+	// resource. Modern composite resources nest all Crossplane machinery fields
+	// under spec.crossplane and status.crossplane, and can't be claimed.
+	SchemaModern Schema = iota
+
+	// SchemaLegacy indicates a LegacyCluster scope composite resource. Legacy
+	// composite resources don't nest Crossplane machinery fields - they're set
+	// directly under spec and status. Legacy composite resources can be claimed.
+	SchemaLegacy
+)
+
+// An Option modifies an unstructured composite resource.
+type Option func(*Unstructured)
+
+// WithGroupVersionKind sets the GroupVersionKind of the composite resource.
+func WithGroupVersionKind(gvk schema.GroupVersionKind) Option {
+	return func(c *Unstructured) {
+		c.SetGroupVersionKind(gvk)
+	}
+}
+
+// WithConditions sets the supplied conditions on the composite resource.
+func WithConditions(c ...xpv1.Condition) Option {
+	return func(cr *Unstructured) {
+		cr.SetConditions(c...)
+	}
+}
+
+// WithSchema sets the schema of the composite resource.
+func WithSchema(s Schema) Option {
+	return func(c *Unstructured) {
+		c.Schema = s
+	}
+}
+
+// New returns a new unstructured composite resource.
+func New(opts ...Option) *Unstructured {
+	c := &Unstructured{Unstructured: unstructured.Unstructured{Object: make(map[string]any)}}
+	for _, f := range opts {
+		f(c)
+	}
+	return c
+}
+
+// +k8s:deepcopy-gen=true
+// +kubebuilder:object:root=true
+
+// An Unstructured composite resource.
+type Unstructured struct {
```

Comment:
Is this a dup of https://github.com/crossplane/crossplane-runtime/blob/main/pkg/resource/unstructured/composite/composite.go ?